File: apps/web/src/lib/api/attendance.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api } from "@/lib/api";
import type {
  AttendanceApproveRequest,
  AttendanceDay,
  AttendanceDayDetail,
  AttendanceDayStatus,
  AttendanceMarkRedRequest,
  AttendanceOverride,
  AttendanceOverrideRequest,
  AttendanceRejectRequest,
  CheckInRequest,
  ISODate,
  Paginated,
  UUID,
} from "@/lib/api/types";

export type AttendanceFilters = {
  assignment_id?: UUID;
  student_id?: UUID;
  status?: AttendanceDayStatus;
  date_from?: ISODate;
  date_to?: ISODate;
};

export const attendanceKeys = {
  all: ["attendance"] as const,
  days: (f: AttendanceFilters, page: number, pageSize: number) =>
    [...attendanceKeys.all, "days", f, page, pageSize] as const,
  day: (id: UUID) => [...attendanceKeys.all, "day", id] as const,
  overrides: (dayId: UUID) => [...attendanceKeys.all, "overrides", dayId] as const,
  today: (assignmentId: UUID) => [...attendanceKeys.all, "today", assignmentId] as const,
};

function qs(filters: AttendanceFilters, page: number, pageSize: number): string {
  const p = new URLSearchParams();
  p.set("page", String(page));
  p.set("page_size", String(pageSize));
  if (filters.assignment_id) p.set("assignment_id", filters.assignment_id);
  if (filters.student_id) p.set("student_id", filters.student_id);
  if (filters.status) p.set("status", filters.status);
  if (filters.date_from) p.set("date_from", filters.date_from);
  if (filters.date_to) p.set("date_to", filters.date_to);
  return p.toString();
}

export function useAttendanceDays(
  filters: AttendanceFilters = {},
  page = 1,
  pageSize = 50,
) {
  return useQuery({
    queryKey: attendanceKeys.days(filters, page, pageSize),
    queryFn: () =>
      api
        .get(`v1/attendance/days?${qs(filters, page, pageSize)}`)
        .json<Paginated<AttendanceDay>>(),
    placeholderData: (prev) => prev,
  });
}

export function useAttendanceDay(id: UUID | null) {
  return useQuery({
    queryKey: id ? attendanceKeys.day(id) : [],
    enabled: !!id,
    queryFn: () => api.get(`v1/attendance/days/${id}`).json<AttendanceDayDetail>(),
  });
}

export function useAttendanceOverrides(dayId: UUID | null) {
  return useQuery({
    queryKey: dayId ? attendanceKeys.overrides(dayId) : [],
    enabled: !!dayId,
    queryFn: () =>
      api.get(`v1/attendance/days/${dayId}/overrides`).json<AttendanceOverride[]>(),
  });
}

export function useMarkRed() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      assignmentId,
      data,
    }: {
      assignmentId: UUID;
      data: AttendanceMarkRedRequest;
    }) =>
      api
        .post(`v1/attendance/assignments/${assignmentId}/mark-red`, { json: data })
        .json<AttendanceDayDetail>(),
    onSuccess: () => qc.invalidateQueries({ queryKey: attendanceKeys.all }),
  });
}

export function useApproveDay() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ id, data }: { id: UUID; data: AttendanceApproveRequest }) =>
      api
        .post(`v1/attendance/days/${id}/approve`, { json: data })
        .json<AttendanceDayDetail>(),
    onSuccess: () => qc.invalidateQueries({ queryKey: attendanceKeys.all }),
  });
}

export function useRejectDay() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ id, data }: { id: UUID; data: AttendanceRejectRequest }) =>
      api
        .post(`v1/attendance/days/${id}/reject`, { json: data })
        .json<AttendanceDayDetail>(),
    onSuccess: () => qc.invalidateQueries({ queryKey: attendanceKeys.all }),
  });
}

export function useOverrideDay() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ id, data }: { id: UUID; data: AttendanceOverrideRequest }) =>
      api
        .post(`v1/attendance/days/${id}/override`, { json: data })
        .json<AttendanceDayDetail>(),
    onSuccess: () => qc.invalidateQueries({ queryKey: attendanceKeys.all }),
  });
}

// Student-only (keyinroq ishlatish uchun)
export function useCheckIn() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      assignmentId,
      data,
    }: {
      assignmentId: UUID;
      data: CheckInRequest;
    }) =>
      api
        .post(`v1/attendance/assignments/${assignmentId}/check-in`, { json: data })
        .json<AttendanceDayDetail>(),
    onSuccess: () => qc.invalidateQueries({ queryKey: attendanceKeys.all }),
  });
}

export function useCheckOut() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      assignmentId,
      data,
    }: {
      assignmentId: UUID;
      data: CheckInRequest;
    }) =>
      api
        .post(`v1/attendance/assignments/${assignmentId}/check-out`, { json: data })
        .json<AttendanceDayDetail>(),
    onSuccess: () => qc.invalidateQueries({ queryKey: attendanceKeys.all }),
  });
}

export function useTodayStatus(assignmentId: UUID | null) {
  return useQuery({
    queryKey: assignmentId ? attendanceKeys.today(assignmentId) : [],
    enabled: !!assignmentId,
    queryFn: () =>
      api
        .get(`v1/attendance/assignments/${assignmentId}/today`)
        .json<AttendanceDayDetail | null>(),
  });
}

File: apps/api/app/api/v1/attendance.py
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Query, Request, status
from sqlalchemy import select

from app.api.deps import (
    CurrentUser,
    RequireAdmin,
    RequireStudent,
    RequireSuperAdmin,
    RequireSupervisor,
)
from app.db.session import SessionDep
from app.models.enums import AttendanceDayStatus, UserRole
from app.models.student import Student
from app.models.supervisor import Supervisor
from app.schemas.attendance import (
    AttendanceApproveRequest,
    AttendanceDayDetail,
    AttendanceDayRead,
    AttendanceMarkRedRequest,
    AttendanceOverrideRead,
    AttendanceOverrideRequest,
    AttendanceRejectRequest,
    CheckInRequest,
    CheckOutRequest,
)
from app.schemas.common import Paginated
from app.services import attendance as svc

router = APIRouter(prefix="/attendance", tags=["attendance"])


# ─── Admin/Supervizor: list + get ────────────────────────


@router.get("/days", response_model=Paginated[AttendanceDayRead])
async def list_days(
    db: SessionDep,
    user: CurrentUser,
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=500),
    assignment_id: UUID | None = None,
    student_id: UUID | None = None,
    status_filter: AttendanceDayStatus | None = Query(None, alias="status"),
    date_from: date | None = None,
    date_to: date | None = None,
    group_id: UUID | None = None,
    direction_id: UUID | None = None,
    faculty_id: UUID | None = None,
) -> Paginated[AttendanceDayRead]:
    # RBAC: Talaba faqat o'z davomatini ko'radi va avto-sync ishga tushadi
    if user.role == UserRole.STUDENT:
        stmt = select(Student.id).where(Student.user_id == user.id)
        current_student_id = (await db.execute(stmt)).scalar_one_or_none()
        if not current_student_id:
            return Paginated(items=[], total=0, page=page, page_size=page_size)
        student_id = current_student_id

    offset = (page - 1) * page_size
    items, total = await svc.list_days(
        db,
        offset,
        page_size,
        assignment_id=assignment_id,
        student_id=student_id,
        status_filter=status_filter,
        date_from=date_from,
        date_to=date_to,
        group_id=group_id,
        direction_id=direction_id,
        faculty_id=faculty_id,
    )
    return Paginated(
        items=[AttendanceDayRead.model_validate(i) for i in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.get("/days/{day_id}", response_model=AttendanceDayDetail)
async def get_day(day_id: UUID, db: SessionDep, _: CurrentUser) -> AttendanceDayDetail:
    return AttendanceDayDetail.model_validate(await svc.get_day(db, day_id))


@router.get("/days/{day_id}/overrides", response_model=list[AttendanceOverrideRead])
async def list_day_overrides(
    day_id: UUID, db: SessionDep, _: RequireAdmin
) -> list[AttendanceOverrideRead]:
    items = await svc.list_overrides(db, day_id)
    return [AttendanceOverrideRead.model_validate(i) for i in items]


# ─── Student: check-in / check-out ──────────────────────


@router.post(
    "/assignments/{assignment_id}/check-in",
    response_model=AttendanceDayDetail,
    status_code=status.HTTP_201_CREATED,
    summary="Talaba: Ishga kelish",
)
async def student_check_in(
    assignment_id: UUID,
    payload: CheckInRequest,
    db: SessionDep,
    user: RequireStudent,
) -> AttendanceDayDetail:
    return AttendanceDayDetail.model_validate(
        await svc.student_check_in(db, assignment_id, user.id, payload)
    )


@router.post(
    "/assignments/{assignment_id}/check-out",
    response_model=AttendanceDayDetail,
    summary="Talaba: Ishdan ketish",
)
async def student_check_out(
    assignment_id: UUID,
    payload: CheckOutRequest,
    db: SessionDep,
    user: RequireStudent,
) -> AttendanceDayDetail:
    return AttendanceDayDetail.model_validate(
        await svc.student_check_out(db, assignment_id, user.id, payload)
    )


@router.get(
    "/assignments/{assignment_id}/today",
    response_model=AttendanceDayDetail | None,
    summary="Talaba: bugungi holat",
)
async def student_today(
    assignment_id: UUID,
    db: SessionDep,
    user: RequireStudent,
) -> AttendanceDayDetail | None:
    data = await svc.student_today_status(db, assignment_id, user.id)
    if not data:
        return None
    return AttendanceDayDetail.model_validate(data)


# ─── Super Admin: approve / reject ────────────────────────────


@router.post(
    "/days/{day_id}/approve",
    response_model=AttendanceDayDetail,
    summary="Super Admin: Yashilga tasdiqlash",
)
async def admin_approve(
    day_id: UUID,
    payload: AttendanceApproveRequest,
    request: Request,
    db: SessionDep,
    user: RequireSuperAdmin,
) -> AttendanceDayDetail:
    from app.services import audit_log as audit

    result = await svc.admin_approve(db, day_id, user.id, payload)
    await audit.log(
        db,
        actor=user,
        action="approve",
        entity_type="attendance_day",
        entity_id=day_id,
        summary="Davomat tasdiqlandi (yashil)",
        metadata={"new_status": "green"},
        request=request,
    )
    await db.commit()
    return AttendanceDayDetail.model_validate(result)


@router.post(
    "/days/{day_id}/reject",
    response_model=AttendanceDayDetail,
    summary="Super Admin: Qizilga rad etish",
)
async def admin_reject(
    day_id: UUID,
    payload: AttendanceRejectRequest,
    request: Request,
    db: SessionDep,
    user: RequireSuperAdmin,
) -> AttendanceDayDetail:
    from app.services import audit_log as audit

    result = await svc.admin_reject(db, day_id, user.id, payload)
    await audit.log(
        db,
        actor=user,
        action="reject",
        entity_type="attendance_day",
        entity_id=day_id,
        summary="Davomat rad etildi (qizil)",
        metadata={"new_status": "red", "note": payload.note},
        request=request,
    )
    await db.commit()
    return AttendanceDayDetail.model_validate(result)


# ─── Super Admin: mark red ───────────────────────────────────


@router.post(
    "/assignments/{assignment_id}/mark-red",
    response_model=AttendanceDayDetail,
    summary="Super Admin: Check-in yo'q kunni qizilga",
)
async def admin_mark_red(
    assignment_id: UUID,
    payload: AttendanceMarkRedRequest,
    request: Request,
    db: SessionDep,
    user: RequireSuperAdmin,
) -> AttendanceDayDetail:
    from app.services import audit_log as audit

    result = await svc.admin_mark_red(db, assignment_id, payload)
    await audit.log(
        db,
        actor=user,
        action="update",
        entity_type="attendance_day",
        entity_id=result.get("id") if isinstance(result, dict) else None,
        summary=f"SuperAdmin davomatni qizil qildi ({payload.date})",
        metadata={
            "new_status": "red",
            "date": str(payload.date),
            "assignment_id": str(assignment_id),
            "note": payload.note,
        },
        request=request,
    )
    await db.commit()
    return AttendanceDayDetail.model_validate(result)


# ─── Super Admin: override ─────────────────────────────


@router.post(
    "/days/{day_id}/override",
    response_model=AttendanceDayDetail,
    summary="Super Admin: qizil ↔ yashil override (sabab majburiy)",
)
async def super_admin_override(
    day_id: UUID,
    payload: AttendanceOverrideRequest,
    request: Request,
    db: SessionDep,
    user: RequireSuperAdmin,
) -> AttendanceDayDetail:
    from app.services import audit_log as audit

    result = await svc.super_admin_override(db, day_id, user.id, payload)
    await audit.log(
        db,
        actor=user,
        action="override",
        entity_type="attendance_day",
        entity_id=day_id,
        summary=f"Davomat override → {payload.new_status.value}",
        metadata={"reason": payload.reason, "new_status": payload.new_status.value},
        request=request,
    )
    await db.commit()
    return AttendanceDayDetail.model_validate(result)
